File: java/device-manager/src/main/java/com/cccode/devicemanager/ontology/service/OntologySchemaService.java
package com.cccode.devicemanager.ontology.service;

import com.cccode.devicemanager.ontology.model.*;
import com.cccode.devicemanager.ontology.repository.*;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.*;

@Service
public class OntologySchemaService {

    private final OntologyObjectTypeRepository objectTypeRepo;
    private final OntologyPropertyRepository propertyRepo;
    private final OntologyLinkTypeRepository linkTypeRepo;

    public OntologySchemaService(OntologyObjectTypeRepository objectTypeRepo,
                                  OntologyPropertyRepository propertyRepo,
                                  OntologyLinkTypeRepository linkTypeRepo) {
        this.objectTypeRepo = objectTypeRepo;
        this.propertyRepo   = propertyRepo;
        this.linkTypeRepo   = linkTypeRepo;
    }

    // ===== ObjectType =====

    public List<OntologyObjectType> listObjectTypes() {
        return objectTypeRepo.findAll();
    }

    public OntologyObjectType getObjectType(Long id) {
        return objectTypeRepo.findById(id)
            .orElseThrow(() -> new NoSuchElementException("ObjectType not found: " + id));
    }

    public OntologyObjectType createObjectType(OntologyObjectType ot) {
        if (objectTypeRepo.existsByName(ot.getName())) {
            throw new IllegalArgumentException("ObjectType name already exists: " + ot.getName());
        }
        return objectTypeRepo.save(ot);
    }

    public OntologyObjectType updateObjectType(Long id, OntologyObjectType patch) {
        OntologyObjectType ot = getObjectType(id);
        if (patch.getDisplayName() != null) ot.setDisplayName(patch.getDisplayName());
        if (patch.getDescription() != null) ot.setDescription(patch.getDescription());
        if (patch.getPrimaryKeyProperty() != null) ot.setPrimaryKeyProperty(patch.getPrimaryKeyProperty());
        if (patch.getIcon() != null) ot.setIcon(patch.getIcon());
        if (patch.getColor() != null) ot.setColor(patch.getColor());
        return objectTypeRepo.save(ot);
    }

    @Transactional
    public void deleteObjectType(Long id) {
        getObjectType(id);
        propertyRepo.deleteByObjectTypeId(id);
        objectTypeRepo.deleteById(id);
    }

    // ===== Property =====

    public List<OntologyProperty> listProperties(Long objectTypeId) {
        return propertyRepo.findByObjectTypeIdOrderBySortOrder(objectTypeId);
    }

    public OntologyProperty addProperty(Long objectTypeId, OntologyProperty prop) {
        getObjectType(objectTypeId); // validate parent exists
        prop.setObjectTypeId(objectTypeId);
        return propertyRepo.save(prop);
    }

    public OntologyProperty updateProperty(Long id, OntologyProperty patch) {
        OntologyProperty prop = propertyRepo.findById(id)
            .orElseThrow(() -> new NoSuchElementException("Property not found: " + id));
        if (patch.getDisplayName() != null) prop.setDisplayName(patch.getDisplayName());
        if (patch.getDataType() != null) prop.setDataType(patch.getDataType());
        if (patch.getDescription() != null) prop.setDescription(patch.getDescription());
        if (patch.getDefaultValue() != null) prop.setDefaultValue(patch.getDefaultValue());
        prop.setRequired(patch.isRequired());
        prop.setSortOrder(patch.getSortOrder());
        return propertyRepo.save(prop);
    }

    public void deleteProperty(Long id) {
        propertyRepo.findById(id)
            .orElseThrow(() -> new NoSuchElementException("Property not found: " + id));
        propertyRepo.deleteById(id);
    }

    // ===== LinkType =====

    public List<OntologyLinkType> listLinkTypes() {
        return linkTypeRepo.findAll();
    }

    public OntologyLinkType getLinkType(Long id) {
        return linkTypeRepo.findById(id)
            .orElseThrow(() -> new NoSuchElementException("LinkType not found: " + id));
    }

    public OntologyLinkType createLinkType(OntologyLinkType lt) {
        getObjectType(lt.getSourceObjectTypeId());
        getObjectType(lt.getTargetObjectTypeId());
        return linkTypeRepo.save(lt);
    }

    public OntologyLinkType updateLinkType(Long id, OntologyLinkType patch) {
        OntologyLinkType lt = getLinkType(id);
        if (patch.getDisplayName() != null) lt.setDisplayName(patch.getDisplayName());
        if (patch.getDescription() != null) lt.setDescription(patch.getDescription());
        if (patch.getCardinality() != null) lt.setCardinality(patch.getCardinality());
        return linkTypeRepo.save(lt);
    }

    public void deleteLinkType(Long id) {
        getLinkType(id);
        linkTypeRepo.deleteById(id);
    }

    // ===== Schema Graph =====

    public Map<String, Object> getSchemaGraph() {
        List<OntologyObjectType> types = objectTypeRepo.findAll();
        List<OntologyLinkType> links   = linkTypeRepo.findAll();

        List<Map<String, Object>> nodes = new ArrayList<>();
        for (OntologyObjectType ot : types) {
            Map<String, Object> node = new LinkedHashMap<>();
            node.put("id",          "ot_" + ot.getId());
            node.put("label",       ot.getDisplayName());
            node.put("name",        ot.getName());
            node.put("type",        "ObjectType");
            node.put("color",       ot.getColor() != null ? ot.getColor() : "#5b8dee");
            node.put("icon",        ot.getIcon());
            node.put("propertyCount", propertyRepo.findByObjectTypeIdOrderBySortOrder(ot.getId()).size());
            nodes.add(node);
        }

        List<Map<String, Object>> edges = new ArrayList<>();
        for (OntologyLinkType lt : links) {
            Map<String, Object> edge = new LinkedHashMap<>();
            edge.put("id",     "lt_" + lt.getId());
            edge.put("source", "ot_" + lt.getSourceObjectTypeId());
            edge.put("target", "ot_" + lt.getTargetObjectTypeId());
            edge.put("label",  lt.getDisplayName());
            edge.put("name",   lt.getName());
            edge.put("cardinality", lt.getCardinality());
            edges.add(edge);
        }

        Map<String, Object> graph = new LinkedHashMap<>();
        graph.put("nodes", nodes);
        graph.put("edges", edges);
        return graph;
    }
}

File: backend/app/services/ontology_schema_service.py
from sqlalchemy.orm import Session
from app.models.ontology import OntologyObjectType, OntologyProperty, OntologyLinkType


# ===== ObjectType =====

def list_object_types(db: Session):
    return db.query(OntologyObjectType).all()


def get_object_type(db: Session, id: int) -> OntologyObjectType:
    ot = db.get(OntologyObjectType, id)
    if not ot:
        raise ValueError(f"ObjectType not found: {id}")
    return ot


def create_object_type(db: Session, data: dict) -> OntologyObjectType:
    existing = db.query(OntologyObjectType).filter(OntologyObjectType.name == data.get("name")).first()
    if existing:
        raise ValueError(f"ObjectType name already exists: {data.get('name')}")
    ot = OntologyObjectType(**data)
    db.add(ot)
    db.commit()
    db.refresh(ot)
    return ot


def update_object_type(db: Session, id: int, data: dict) -> OntologyObjectType:
    ot = get_object_type(db, id)
    for key, value in data.items():
        if value is not None:
            setattr(ot, key, value)
    db.commit()
    db.refresh(ot)
    return ot


def delete_object_type(db: Session, id: int):
    get_object_type(db, id)
    db.query(OntologyProperty).filter(OntologyProperty.object_type_id == id).delete()
    db.query(OntologyObjectType).filter(OntologyObjectType.id == id).delete()
    db.commit()


# ===== Property =====

def list_properties(db: Session, object_type_id: int):
    return db.query(OntologyProperty).filter(
        OntologyProperty.object_type_id == object_type_id
    ).order_by(OntologyProperty.sort_order).all()


def add_property(db: Session, object_type_id: int, data: dict) -> OntologyProperty:
    get_object_type(db, object_type_id)
    prop = OntologyProperty(object_type_id=object_type_id, **data)
    db.add(prop)
    db.commit()
    db.refresh(prop)
    return prop


def update_property(db: Session, id: int, data: dict) -> OntologyProperty:
    prop = db.get(OntologyProperty, id)
    if not prop:
        raise ValueError(f"Property not found: {id}")
    for key, value in data.items():
        if value is not None or key in ("required", "sort_order"):
            setattr(prop, key, value)
    db.commit()
    db.refresh(prop)
    return prop


def delete_property(db: Session, id: int):
    prop = db.get(OntologyProperty, id)
    if not prop:
        raise ValueError(f"Property not found: {id}")
    db.delete(prop)
    db.commit()


# ===== LinkType =====

def list_link_types(db: Session):
    return db.query(OntologyLinkType).all()


def get_link_type(db: Session, id: int) -> OntologyLinkType:
    lt = db.get(OntologyLinkType, id)
    if not lt:
        raise ValueError(f"LinkType not found: {id}")
    return lt


def create_link_type(db: Session, data: dict) -> OntologyLinkType:
    get_object_type(db, data["source_object_type_id"])
    get_object_type(db, data["target_object_type_id"])
    lt = OntologyLinkType(**data)
    db.add(lt)
    db.commit()
    db.refresh(lt)
    return lt


def update_link_type(db: Session, id: int, data: dict) -> OntologyLinkType:
    lt = get_link_type(db, id)
    for key, value in data.items():
        if value is not None:
            setattr(lt, key, value)
    db.commit()
    db.refresh(lt)
    return lt


def delete_link_type(db: Session, id: int):
    get_link_type(db, id)
    db.query(OntologyLinkType).filter(OntologyLinkType.id == id).delete()
    db.commit()


# ===== Schema Graph =====

def get_schema_graph(db: Session) -> dict:
    types = db.query(OntologyObjectType).all()
    links = db.query(OntologyLinkType).all()

    nodes = []
    for ot in types:
        prop_count = db.query(OntologyProperty).filter(
            OntologyProperty.object_type_id == ot.id
        ).count()
        nodes.append({
            "id": f"ot_{ot.id}",
            "label": ot.display_name,
            "name": ot.name,
            "type": "ObjectType",
            "color": ot.color or "#5b8dee",
            "icon": ot.icon,
            "propertyCount": prop_count,
        })

    edges = []
    for lt in links:
        edges.append({
            "id": f"lt_{lt.id}",
            "source": f"ot_{lt.source_object_type_id}",
            "target": f"ot_{lt.target_object_type_id}",
            "label": lt.display_name,
            "name": lt.name,
            "cardinality": lt.cardinality,
        })

    return {"nodes": nodes, "edges": edges}
